src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509chain.c

```diff
@@ -0,0 +1,598 @@
+    BASICREASON_REVOKED,
+    BASICREASON_UNDETERMINED_REVOCATION_STATUS,
+    BASICREASON_INVALID_SIGNATURE,
+    BASICREASON_ALGORITHM_CONSTRAINED,
+};
+
+static PAL_X509ChainStatusFlags ChainStatusFromValidatorExceptionReason(JNIEnv* env, jobject reason)
+{
+    int value = (*env)->CallIntMethod(env, reason, g_EnumOrdinal);
+    if (g_CertPathExceptionBasicReasonClass != NULL
+        && (*env)->IsInstanceOf(env, reason, g_CertPathExceptionBasicReasonClass))
+    {
+        switch (value)
+        {
+            case BASICREASON_UNSPECIFIED:
+                return PAL_X509ChainPartialChain;
+            case BASICREASON_EXPIRED:
+            case BASICREASON_NOT_YET_VALID:
+                return PAL_X509ChainNotTimeValid;
+            case BASICREASON_REVOKED:
+                return PAL_X509ChainRevoked;
+            case BASICREASON_UNDETERMINED_REVOCATION_STATUS:
+                return PAL_X509ChainRevocationStatusUnknown;
+            case BASICREASON_INVALID_SIGNATURE:
+                return PAL_X509ChainNotSignatureValid;
+            case BASICREASON_ALGORITHM_CONSTRAINED:
+                return PAL_X509ChainPartialChain;
+        }
+    }
+    else if (g_PKIXReasonClass != NULL && (*env)->IsInstanceOf(env, reason, g_PKIXReasonClass))
+    {
+        switch (value)
+        {
+            case PKIXREASON_NAME_CHAINING:
+                return PAL_X509ChainPartialChain;
+            case PKIXREASON_INVALID_KEY_USAGE:
+                return PAL_X509ChainNotValidForUsage;
+            case PKIXREASON_INVALID_POLICY:
+                return PAL_X509ChainInvalidPolicyConstraints;
+            case PKIXREASON_NO_TRUST_ANCHOR:
+                return PAL_X509ChainPartialChain;
+            case PKIXREASON_UNRECOGNIZED_CRIT_EXT:
+                return PAL_X509ChainHasNotSupportedCriticalExtension;
+            case PKIXREASON_NOT_CA_CERT:
+                return PAL_X509ChainUntrustedRoot;
+            case PKIXREASON_PATH_TOO_LONG:
+                return PAL_X509ChainInvalidBasicConstraints;
+            case PKIXREASON_INVALID_NAME:
+                return PAL_X509ChainInvalidNameConstraints;
+        }
+    }
+
+    return PAL_X509ChainPartialChain;
+}
+
+static void PopulateValidationError(JNIEnv* env, jobject error, bool isRevocationError, ValidationError* out)
+{
+    int index = -1;
+    PAL_X509ChainStatusFlags chainStatus = PAL_X509ChainNoError;
+    if ((*env)->IsInstanceOf(env, error, g_CertPathValidatorExceptionClass))
+    {
+        index = (*env)->CallIntMethod(env, error, g_CertPathValidatorExceptionGetIndex);
+
+        // Get the reason (if the API is available) and convert it to a chain status flag
+        if (g_CertPathValidatorExceptionGetReason != NULL)
+        {
+            jobject reason = (*env)->CallObjectMethod(env, error, g_CertPathValidatorExceptionGetReason);
+            chainStatus = ChainStatusFromValidatorExceptionReason(env, reason);
+            (*env)->DeleteLocalRef(env, reason);
+        }
```

Comment:
Shouldn't this `else` set chainStatus to... something?

We have a validation exception, but can't ask what went wrong... that's not "NoError" :)

Reply:
Err, yep. Apparently, have the managed just forcing `NoError` -> `PartialChain` for... reasons? Will make this fallback, since that makes way more sense.